src/backend/distributed/commands/index.c

```diff
@@ -172,65 +173,94 @@ PlanIndexStmt(IndexStmt *createIndexStatement, const char *createIndexCommand)
 
 
 /*
- * ErrorIfReindexOnDistributedTable determines whether a given REINDEX
- * involves a distributed table, & raises an error if so.
+ * PlanReindexStmt determines whether a given REINDEX statement involves
+ * a distributed table. If so (and if the statement does not use unsupported
+ * options), it modifies the input statement to ensure proper execution against
+ * the master node table and creates a DDLJob to encapsulate information needed
+ * during the worker node portion of DDL execution before returning that DDLJob
+ * in a List. If no distributed table is involved, this function returns NIL.
  */
-void
-ErrorIfReindexOnDistributedTable(ReindexStmt *ReindexStatement)
+List *
+PlanReindexStmt(ReindexStmt *ReindexStatement, const char *ReindexCommand)
 {
-	Relation relation = NULL;
-	Oid relationId = InvalidOid;
-	bool isDistributedRelation = false;
-	LOCKMODE lockmode = AccessShareLock;
+	List *ddlJobs = NIL;
 
 	/*
 	 * We first check whether a distributed relation is affected. For that, we need to
-	 * open the relation.
+	 * open the relation. To prevent race conditions with later lookups, lock the table,
+	 * and modify the rangevar to include the schema.
 	 */
-	if (ReindexStatement->relation == NULL)
+	if (ReindexStatement->relation != NULL)
 	{
-		/* ignore REINDEX SCHEMA, REINDEX SYSTEM, and REINDEX DATABASE */
-		return;
-	}
+		Relation relation = NULL;
+		Oid relationId = InvalidOid;
+		bool isDistributedRelation = false;
+		char *namespaceName = NULL;
+#if PG_VERSION_NUM >= 120000
+		LOCKMODE lockmode = ReindexStatement->concurrent ? ShareUpdateExclusiveLock :
+							AccessExclusiveLock;
+#else
+		LOCKMODE lockmode = AccessExclusiveLock;
+#endif
+		MemoryContext relationContext = NULL;
 
-	Assert(ReindexStatement->kind == REINDEX_OBJECT_INDEX ||
-		   ReindexStatement->kind == REINDEX_OBJECT_TABLE);
+		/*
+		 * XXX: Consider using RangeVarGetRelidExtended with a permission
+		 * checking callback. Right now we'll acquire the lock before having
+		 * checked permissions, and will only fail when executing the actual
+		 * index statements.
+		 */
+		if (ReindexStatement->kind == REINDEX_OBJECT_INDEX)
+		{
+			Oid indOid = RangeVarGetRelid(ReindexStatement->relation,
+										  NoLock, false);
```

Comment:
Could we copy-paste `RangeVarCallbackForReindexIndex` from postgres and use `RangeVarGetRelidExtended ` here?
https://github.com/postgres/postgres/blob/master/src/backend/commands/indexcmds.c#L2329

The reason that Postgres does the permissions check so early is that it would otherwise allow arbitrary users to get in the queue for an access exclusive lock. That means that if there is a long-running SELECT (possibly initiated by the user himself), all other queries that use the index will block. To prevent arbitrary users from completely messing up the database, aggressive locks are guarded with `RangeVarGetRelidExtended`.

We don't always do this in Citus, but we should.

Reply:
Definitely lower priority than procedure call delegation tasks, though it seems small and close so if you have some time remaining (e.g. while waiting for review or at the end of the release cycle), this would be a nice thing to merge.

Reply:
Updated to use `RangeVarGetRelidExtended` for `REINDEX INDEX` & `EnsureTableOwner` for `REINDEX TABLE`